src/core/ir/discriminator.rs

```diff
@@ -0,0 +1,1121 @@
+            }
+
+            if !duplicates.is_empty() {
+                return Valid::from_vec_cause(
+                    duplicates
+                        .into_iter()
+                        .map(|(_, same_types)| {
+                            Cause::new(format!(
+                                "Union have equal types: {} ",
+                                same_types.join(" == ")
+                            ))
+                        })
+                        .collect(),
+                )
+                .trace(union_name);
+            }
+        }
+
+        // strip some fields that are not valuable for discriminator
+        let fields_info = {
+            let mut seen_required_in: HashSet<usize> = HashSet::new();
+
+            fields_info
+                .into_iter()
+                .filter(|(_, field_info)| {
+                    let drop =
+                        // if field is presented in all types then it doesn't help in figuring out the type of value
+                        field_info
+                        .presented_in
+                        .is_covering_all_types(union_types.len())
+                        // if multiple fields are required in the same set of types than we can leave only one of such fields
+                        || (*field_info.required_in != 0 && seen_required_in.contains(&field_info.required_in));
+
+                    seen_required_in.insert(*field_info.required_in);
+
+                    !drop
+                })
+                .collect()
+        };
+
+        let discriminator = Self { fields_info, types };
+
+        tracing::debug!(
+            "Generated discriminator for union type '{union_name}':\n{discriminator:?}",
+        );
+
+        Valid::succeed(discriminator)
+    }
+
+    pub fn resolve_type(&self, value: &Value) -> Result<TypeName> {
+        if let Value::List(list) = value {
+            let results: Result<Vec<_>> = list
+                .iter()
+                .map(|item| Ok(self.resolve_type_for_single(item)?.to_string()))
+                .collect();
+
+            Ok(TypeName::Vec(results?))
+        } else {
+            Ok(TypeName::Single(
+                self.resolve_type_for_single(value)?.to_string(),
+            ))
+        }
+    }
+
+    fn resolve_type_for_single(&self, value: &Value) -> Result<&str> {
+        let Value::Object(obj) = value else {
+            bail!("Value expected to be object");
+        };
+
+        let mut possible_types = BitVec::set_after(self.types.len());
+
+        for (field, info) in &self.fields_info {
+            if obj.contains_key(field.as_str()) {
+                *possible_types &= *info.presented_in;
+            } else {
+                *possible_types &= !*info.required_in;
+            }
+
+            if possible_types.is_empty() {
+                // no possible types, something is wrong with the resolved value
+                bail!("Failed to find corresponding type for value")
```

Comment:
integration tests for all possible bailout conditions.

Reply:
added tests